core/drivers/firewall/stm32_risab.c

```diff
@@ -0,0 +1,843 @@
+#define _RISAB_IASR				U(0x8)
+#define _RISAB_IACR				U(0xC)
+#define _RISAB_RCFGLOCKR			U(0x10)
+#define _RISAB_IAESR				U(0x20)
+#define _RISAB_IADDR				U(0x24)
+#define _RISAB_PGy_SECCFGR(y)			(U(0x100) + (0x4 * (y)))
+#define _RISAB_PGy_PRIVCFGR(y)			(U(0x200) + (0x4 * (y)))
+#define _RISAB_RISAB_PGy_C2PRIVCFGR(y)		(U(0x600) + (0x4 * (y)))
+#define _RISAB_CIDxPRIVCFGR(x)			(U(0x800) + (0x20 * (x)))
+#define _RISAB_CIDxRDCFGR(x)			(U(0x808) + (0x20 * (x)))
+#define _RISAB_CIDxWRCFGR(x)			(U(0x810) + (0x20 * (x)))
+#define _RISAB_PGy_CIDCFGR(y)			(U(0xA00) + (0x4 * (y)))
+#define _RISAB_HWCFGR3				U(0xFE8)
+#define _RISAB_HWCFGR2				U(0xFEC)
+#define _RISAB_HWCFGR1				U(0xFF0)
+#define _RISAB_VERR				U(0xFF4)
+#define _RISAB_IPIDR				U(0xFF8)
+#define _RISAB_SIDR				U(0xFFC)
+
+/* RISAB_CR bitfields */
+#define _RISAB_CR_SRWIAD			BIT(31)
+
+/* RISAB_IACR bitfields */
+#define _RISAB_IACR_CAEF			BIT(0)
+#define _RISAB_IACR_IAEF			BIT(1)
+
+/* Define RISAB_PG_SECCFGR bitfields */
+#define _RISAB_PG_SECCFGR_MASK			GENMASK_32(7, 0)
+
+/* Define RISAB_PG_PRIVCFGR bitfields */
+#define _RISAB_PG_PRIVCFGR_MASK			GENMASK_32(7, 0)
+
+/* CIDCFGR bitfields */
+#define _RISAB_PG_CIDCFGR_CFEN			BIT(0)
+#define _RISAB_PG_CIDCFGR_DCEN			BIT(2)
+#define _RISAB_PG_CIDCFGR_DDCID_SHIFT		U(4)
+#define _RISAB_PG_CIDCFGR_DDCID_MASK		GENMASK_32(6, 4)
+#define _RISAB_PG_CIDCFGR_CONF_MASK		(_RISAB_PG_CIDCFGR_CFEN | \
+						 _RISAB_PG_CIDCFGR_DCEN | \
+						 _RISAB_PG_CIDCFGR_DDCID_MASK)
+
+/* Miscellaneous */
+#define _RISAB_NB_PAGES_MAX			U(32)
+
+#define _RISAB_PAGE_SIZE			U(0x1000)
+
+#define _RISAB_NB_MAX_CID_SUPPORTED		U(7)
+
+struct mem_region {
+	paddr_t base;
+	size_t size;
+};
+
+struct stm32_risab_rif_conf {
+	unsigned int first_page;
+	unsigned int nb_pages_cfged;
+	uint32_t plist[_RISAB_NB_MAX_CID_SUPPORTED];
+	uint32_t rlist[_RISAB_NB_MAX_CID_SUPPORTED];
+	uint32_t wlist[_RISAB_NB_MAX_CID_SUPPORTED];
+	uint32_t cidcfgr;
+	uint32_t dprivcfgr;
+	uint32_t seccfgr;
+};
+
+struct stm32_risab_pdata {
+	unsigned int nb_regions_cfged;
+	struct clk *clock;
+	struct mem_region region_cfged;
+	struct stm32_risab_rif_conf *subr_cfg;
+	struct io_pa_va base;
+	unsigned int conf_lock;
+	char risab_name[20];
+	uint32_t pages_configured;
+	bool srwiad;
+
+	SLIST_ENTRY(stm32_risab_pdata) link;
+};
+
+static SLIST_HEAD(, stm32_risab_pdata) risab_list =
+		SLIST_HEAD_INITIALIZER(risab_list);
+
+static bool is_tdcid;
+
+static vaddr_t risab_base(struct stm32_risab_pdata *risab)
+{
+	return io_pa_or_va_secure(&risab->base, 1);
+}
+
+void stm32_risab_clear_illegal_access_flags(void)
+{
+	struct stm32_risab_pdata *risab = NULL;
+
+	SLIST_FOREACH(risab, &risab_list, link) {
+		vaddr_t base = risab_base(risab);
+
+		if (!io_read32(base + _RISAB_IASR))
+			continue;
+
+		io_write32(base + _RISAB_IACR, _RISAB_IACR_CAEF |
+			   _RISAB_IACR_IAEF);
+	}
+}
+
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risab_print_erroneous_data(void)
+{
+	struct stm32_risab_pdata *risab = NULL;
+
```

Comment:
I suggest to remove the `#ifndef` directive at line 129 and add here
```c
	if (!IS_ENABLED(CFG_TEE_CORE_DEBUG))
		return;
```
to be consistent with inline description comment in `stm32_risab.h`.

Reply:
I'll prefer to stub. It will be used outside the file. Therefore, I'd prefer to avoid compile issues if CFG_STM32_RISAB is not set

Reply:
Then header file should use `if defined(CFG_STM32_RISAB) && defined(CFG_TEE_CORE_DEBUG)`.

Maybe consider stubbing also `stm32_risab_clear_illegal_access_flags()` in the header file `!defined(CFG_STM32_RISAB)`.